mobilecoind/src/service.rs

```diff
@@ -4184,6 +4306,249 @@ mod test {
+                    assert_eq!(amount.value, offered_value / 4 - Mob::MINIMUM_FEE);
+                    found_counterparty_output = true;
+                } else {
+                    assert_eq!(*amount.token_id, 1);
+                    // The quote asks for 123 eusd, counterparty fulfills to 1/4,
+                    // rounding up so they give away 31.
+                    // Change is their eusd utxo value minus that.
+                    assert_eq!(amount.value, counterparty_eusd_utxo_value - 31);
+                    found_counterparty_change = true;
+                }
+            } else if let Ok((amount, _scalar)) =
+                output.view_key_match(swap_originator.view_private_key())
+            {
+                if amount.token_id == Mob::ID {
+                    assert_eq!(amount.value, offered_value * 3 / 4);
+                    found_originator_change = true;
+                } else {
+                    // 31 is 1/4 of 123, rounded up
+                    assert_eq!(amount, Amount::new(31, TokenId::from(1)));
+                    found_originator_output = true;
+                }
+            }
+        }
+        assert!(found_counterparty_output);
+        assert!(found_counterparty_change);
+        assert!(found_originator_output);
+        assert!(found_originator_change);
+
+        // Changing the fee token id to 1 should work, and slightly adjust the output
+        // values.
+        request.set_fee_token_id(1);
+
+        let generate_mixed_tx_response = client.generate_mixed_tx(&request).unwrap();
+        let tx = Tx::try_from(generate_mixed_tx_response.get_tx_proposal().get_tx()).unwrap();
+
+        assert_eq!(tx.prefix.outputs.len(), 4);
+
+        let mut found_counterparty_output = false;
+        let mut found_counterparty_change = false;
+        let mut found_originator_output = false;
+        let mut found_originator_change = false;
+        for output in tx.prefix.outputs.iter() {
+            if let Ok((amount, _scalar)) =
+                output.view_key_match(swap_counterparty.view_private_key())
+            {
+                if amount.token_id == Mob::ID {
+                    assert_eq!(amount.value, offered_value / 4);
+                    found_counterparty_output = true;
+                } else {
+                    assert_eq!(*amount.token_id, 1);
+                    // The quote asks for 123 eusd, counterparty fulfills to 1/4,
+                    // rounding up so they give away 31.
+                    // Change is their eusd utxo value minus that, minus the eusd transaction fee.
+                    // FIXME: I'm not sure where the 10 billion number comes from, it is not the
+                    // same as the value of Mob::MINIMUM_FEE, and doesn't appear to be in
+                    // FeeMap::default()
+                    assert_eq!(
+                        amount.value,
+                        counterparty_eusd_utxo_value - 31 - 10_000_000_000
```

Comment:
this is the fee it's actually selecting right now, with the default map...

Reply:
Looks like there's a `FALLBACK_FEE` const...

Reply:
seems questionable to me to allow this, it could result in people paying much higher fees than expected if something is misconfigured

Reply:
okay, i got rid of fallback fee and extended the testing fee map as you suggested in the latest commit, and there are more tests of what happens when the user specifies a fee value